content/blog/2019-09-25-vril-vector-bril.md

```diff
@@ -0,0 +1,139 @@
+We first wrote a simple test to initialize an array, store a value to an array and read a value from an array. 
+A second test described a loop which would add two array elements and save it in a third array.
+Both these tests generated the expected result from the interpreter. We also walked through the program to count the dynamic instruction count and number of hops over basic blocks. A program with 50 lines of code created 109 dynamic instructions and 11 hops, which matched the expectation. These tests reside in `vtest/interp` directory.
+
+### Evaluating Vector operation `vadd`
+We took a similar approach to evaluate our addition of `vadd` in Vril.
+- evaluating integration to Bril
+- evaluating functionality of `vadd`
+- evaluating performance of `vadd`
+
+#### Integration tests for Vectors
+We added tests to generate JSON and text representation of Bril from each other to gain confidence about its translatability. These tests are included in `vtest/parse` and `vtest/print`.
+
+#### Functional testing of `vadd`
+We wrote a program with `vadd` and ran this test in `vrili` with different values for `maxvl` which represents the size of vector lenght of the backend. The test generated expected results. This is included in `vtest/interp/array_vector.bril`.
+
+#### Performance of `vadd`
+Initial motivation for Vril is to reduce hop counts by having vector operators in Bril. This aspect is tested by comparing dynamic instruction count, hop count and lines of code for a loop executed in a scalar manner and a vector operation. 
+
+| Program           | Lines of code | Dynamic instructions | Hop count |
+|-------------------|---------------|----------------------|-----------|
+| Conventional Bril | 43            | 41                   | 1         |
+| Scalar Array      | 50            | 109                  | 12        |
+| Vector Array      | 48            | 44                   | 3         |
+
+Given that the length of array is N=10 and vector length selected by Vrili at execution is k, and compensating for the start and the end block as shown in the figure, scalar operation iterates N times and vector operation iterates N/k times.
+
+CFG for scalar array operation  | CFG for vector array operation
+:-------------------------:|:-------------------------:
+![](https://github.com/sa2257/vril/blob/82e9da0eed50a98fc5285a32020b0ba59385e398/plots/array_scalar_nv.png) | ![](https://github.com/sa2257/vril/blob/82e9da0eed50a98fc5285a32020b0ba59385e398/plots/array_vector_nv.png)
+
+Conclusion
+--------------------------------------
+We have extended bril to support array types. We have added two new operations to move data in and out of the arrays so that we can emulate data movements between an array and what it would be a scalar register. We have also extended bril to support vector operations of two types: configuration and arithmetic. Configuration operations allow to modify a vector state and arithmetic operations perform operations on array arguments.
```

Comment:
```suggestion
We have extended bril to support array types. We have added two new operations to move data in and out of the arrays so that we can emulate data movements between an array and what it would be a scalar register. We have also extended bril to support vector operations of two types: configuration and arithmetic. Configuration operations allow programs to modify a vector state and arithmetic operations perform operations on array arguments.
```
(Missing subject.)